databroker/src/grpc/kuksa_val_v2/val.rs

```diff
@@ -27,44 +28,92 @@ use databroker_proto::kuksa::val::v2::{
     open_provider_stream_response, OpenProviderStreamResponse, PublishValuesResponse,
 };
 
-use kuksa::proto::v2::{ListMetadataResponse, Metadata};
+use kuksa::proto::v2::{
+    signal_id, ActuateRequest, ActuateResponse, BatchActuateStreamRequest, ListMetadataResponse,
+    Metadata, ProvideActuationResponse,
+};
 use std::collections::HashSet;
 use tokio::{select, sync::mpsc};
 use tokio_stream::{wrappers::ReceiverStream, Stream, StreamExt};
 use tracing::debug;
 
 const MAX_REQUEST_PATH_LENGTH: usize = 1000;
 
+pub struct Provider {
+    sender: mpsc::Sender<Result<OpenProviderStreamResponse, tonic::Status>>,
+}
+
+#[async_trait::async_trait]
+impl ActuationProvider for Provider {
+    async fn actuate(
+        &self,
+        actuation_changes: Vec<broker::ActuationChange>,
+    ) -> Result<(), (broker::ActuationError, String)> {
+        let mut actuation_requests: Vec<ActuateRequest> = vec![];
+        for actuation_change in actuation_changes {
+            let data_value = actuation_change.data_value;
+            actuation_requests.push(ActuateRequest {
+                signal_id: Some(proto::SignalId {
+                    signal: Some(signal_id::Signal::Id(actuation_change.id)),
+                }),
+                value: Some(proto::Value::from(data_value)),
+            });
+        }
+
+        let batch_actuate_stream_request =
+            open_provider_stream_response::Action::BatchActuateStreamRequest(
+                BatchActuateStreamRequest {
+                    actuate_requests: actuation_requests,
+                },
+            );
+
+        let response = OpenProviderStreamResponse {
+            action: Some(batch_actuate_stream_request),
+        };
+
+        let result = self.sender.send(Ok(response)).await;
+        if result.is_err() {
+            let send_error = result.unwrap_err().0;
+            if send_error.is_err() {
```

Comment:
This if-statement is quite useless as we anyway return an error after, or?

Reply:
Yes, you are right